Improved the time complexity of TensoredFilter.apply() into O(n * 2^n)

BOBO1997 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

This PR improves the time complexity of qiskit.ignis.mitigation.TensoredFilter.apply() from O(n * 4^n) to O(n * 2^n).

### Details and comments

The apply() method of TensoredFilter class in Ignis includes two options of implementation: "pseudo_inverse" and "least_squares".
Currently, both options cost O(n * 4^n) time to the qubit size n, because the two nested for loops with 2^n iterations were used in the multiplication of calibration matrices (or the pseudo-inverse of them) to the probability vector.

I fixed the time complexity of this part into O(n * 2^n), keeping the outcome of this part exactly the same as the previous code.
For example, this improvement allows us to implement the tensored mitigation of the 6 qubit system with the "least_squares" method in a few seconds, while the previous code costs almost 1 minute (in my environment).

### Tests to cover my changes

The brief benchmarking tests using GHZ state are attached to this PR.
We can see the improvement in the calculation time and we can also see the fixed code returns the almost same result for the mitigation counts. (I'm sorry that I didn't use the same random seed in the following experiments.)

- [tensored_filter_fixed_least_squares.pdf](https://github.com/Qiskit/qiskit-ignis/files/5878070/tensored_filter_fixed_least_squares.pdf) tests the proposed code for "least square" method
- [tensored_filter_original_least_squares.pdf](https://github.com/Qiskit/qiskit-ignis/files/5878072/tensored_filter_original_least_squares.pdf) tests the previous code for "least square" method

- [tensored_filter_fixed_pseudo_inverse.pdf](https://github.com/Qiskit/qiskit-ignis/files/5878071/tensored_filter_fixed_pseudo_inverse.pdf) tests the proposed code for "pseudo inverse" method
- [tensored_filter_original_pseudo_inverse.pdf](https://github.com/Qiskit/qiskit-ignis/files/5878073/tensored_filter_original_pseudo_inverse.pdf) tests the previous code for "pseudo inverse" method

Since this PR is a modification of the internal behavior of the TensoredFilter.apply() function, I did not change documents.
Users can still use this function in the same way as before.

### Limitations

Although the efficiency is slightly improved, the time complexity is still in the exponential scale to the qubit size.
One of the solutions to the exponential resource requirement can be found in https://arxiv.org/abs/2101.08946 by Mooney et al. which cuts off the elements with neglectable small values in the probability vector.
Implementing their approach may require changing whole parts of the TensoredFilter.apply() function since this function essentially creating a 2^n length probability vector.

Thank you very much for paying attention to this PR!
Please point out any mistakes in this PR.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-ignis?pullRequest=551) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-ignis?pullRequest=551) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-ignis?pullRequest=551) it.</sub>

yaelbh commented:
Thanks @BOBO1997 !    Can you please check why a test is failing: https://github.com/Qiskit/qiskit-ignis/pull/551/checks?check_run_id=1774987466.   You can run tests locally in your environment by `python -m unittest`.

BOBO1997 commented:
Dear Dr. Ben-Haim, thank you so much for your advice!
I found my code has not met all the test cases, especially in the case of using partial complete matrices such as mit_pattern = [[2], [4,1]].
I am now fixing this problem.

Through further investigation, I also found that the specific position of each qubit that calibration matrices cover was ignored in the process of TensoredFilter.apply().
This caused the following problem:
If mit_pattern is [[1, 3], [0], [2, 4]] and the noisy qubit is qubit 0, then the TensoredFilter.apply() would mitigate the error on qubit 1 using the calibration matrix of qubit 0, which seems to be unexpected behavior.
This kind of bug cannot be detected by current unittest where the same errors are used for each qubit.

I think I can fix this part using "mit_pattern" in the TensoredFilter.apply() function.
In this sense, several variables and functions would no longer be needed in TensoredFilter.
- self._qubit_list_sizes
- self._indices_list
- self.substate_labels_list()

If these bugs have already discussed before, then please don't worry about my comments and I'm sorry for repeating the same discussion.

### WIP
So far, another bug occurred by using mit_pattern in TensoredFilter.apply().
To fix the current bug, I guess I have to use the information of QuantumRegister-ClassicalRegister mapping in the measurement process.
The specification of TensoredMeasFitter would be slightly changed (←adding a new variable representing such mapping).
I would like to add a new commit after fixing whole bugs in my code.

### TODO
The remaining tasks are listed below, in order to solve whole errors.

- add a new variable to TensoredMeasFitter to fix the current bug
- add test case for different readout errors in different qubits
- use integers to represent quantum states (enhancement)

yaelbh commented:
@mtreinish The CI items are "expected" - what does it mean, and what's preventing them from becoming "successful" ?

mtreinish commented:
Expected is just the status for jobs that are required but have not started running yet. It looks like there was a github actions outage this morning: https://www.githubstatus.com/incidents/tf9v5jjmq2lg which probably preventedtthe jobs from being triggered.

yaelbh commented:
Is there a way to retrigger them, other than pushing an empty commit?

BOBO1997 commented:
@yaelbh @mtreinish  Thank you for the information about CI processing, the CI processing seems to have recovered now.

Regarding the current fails in several tests, all of them show the assertion error in 
[L.483 of test_meas.py](https://github.com/BOBO1997/qiskit-ignis/blob/accelerate_tensored_filter/test/measurement_calibration/test_meas.py#L483,) where they compare the following two counts.

output_results_pseudo_inverse (tested counts)
{'000': 482.86648558743315,
 '001': 8.071748430570736,
 '010': 4.919896570839651,
 '011': -9.98085488627964,
 '100': 17.788244925765085,
 '101': -18.267077974735734,
 '110': -6.777514012666714,
 '111': 521.3790713590731}

saved_info['results_pseudo_inverse'] (expected counts)
{'000': 490.09920463808,
 '001': 1.1855217632343553,
 '010': 6.856561953208232,
 '011': -13.925536339030689,
 '100': 21.080379879311838,
 '101': -19.929567717733956,
 '110': 2.4129313160877572,
 '111': 512.2205045068418}

The tested counts is mitigated by [TensoredMeasFitter with substate_labels_list](https://github.com/BOBO1997/qiskit-ignis/blob/accelerate_tensored_filter/test/measurement_calibration/test_meas.py#L459).
Since the test cases examined the counts of GHZ state, the tested counts seems to be a correct distribution while it did not exactly match the expected one.

I would like to clarify that:
- Is it necessary for the tested counts to exactly correspond to the expected counts which is the mitigated counts using the  TensoredMeasFitter **without** substate_labels_list?

yaelbh commented:
I don't see a reason for this to be necessary (by the way thanks for the nice work you're doing here 😃 )

BOBO1997 commented:
Thank you for your encouraging comment!

I found the previous fail test cases are from my misunderstanding of some variable in TensoredFilter.
The latest commit passes these tests.

### WIP
There are still some fail cases.

1. the dependency conflict (←packages such as cvxpy could not be installed. Maybe we should modify tox.ini ??)
2. docs build error (←mainly related to the deprecation of `np.int`)
3. CellTimeoutError in tutorials/noise/5_quantum_volume.ipynb

For future tasks, I would fix these failings as much as possible.
Since I only modified the code related to the algorithm part for error mitigation, these failings are totally unexpected to me.
If these are the common cases, I would be grateful if you could share some tips.

yaelbh commented:
Indeed, this happened also to me yesterday, I received a mail stating that my master branch in Ignis is failing the CI tests. @mtreinish will you have time to help here?

ShellyGarion commented:
I think that this PR in Terra should fix some of the problems:
https://github.com/Qiskit/qiskit-terra/pull/5758

yaelbh commented:
@BOBO1997 No action item from you right now. Let's wait till the CI stabilizes (master versions of Aer and Ignis are not passing currently, probably due to Terra). Then, when we have green CI in this pull request, we'll be able to finish reviewing it.

BOBO1997 commented:
Thank you for your support! Then I'm waiting for the all green CI state.

yaelbh commented:
Although, admittedly, the original code is not well-documented. Could you please document:
- The general idea of how the calculation is done (for example, a mathematical formula and the steps to calculate it, along with pointers to where each step is done in the code, and how parts of the formula are expressed as variables in the code).
- Explanation of what most lines do.
- A dry run of a small and simple example.

yaelbh commented:
A couple of general comments about the topic:
1. Regarding  https://arxiv.org/abs/2101.08946: a suggestion for performance improvement of tensored mitigation is given towards the bottom of the first column of Page 4 (right before "Parity Verification"). This seems to be a good heuristic, and indeed, the authors claim that it works well for them, both in run time and in accuracy. However we should still bare in mind that (at least as far as I see) at least the first step in the heuristic (application of the first inverse calibration matrix) still seems exponential; and, more importantly, there are no guarantees about the accuracy of the approximation that's done in the heuristic.
2. You may be interested in the discussion in #547.

BOBO1997 commented:
Okay, I added some words to the docs for TensoredFilter.apply().
Are they specific enough, or do I have to explain in more detail?

### Errors in Tests / docs
After I added documents to TensoredFilter.apply() and pushed the changes to the remote, the tests for docs failed.
It seems that the build of documents fails due to some errors (directly, InvocationError of tox command).
I also checked the previous commits still fails in my environment (but they [passed all the tests](https://github.com/Qiskit/qiskit-ignis/pull/551/commits/3c4380a1ba6d611110e57e40f227d9d3a65b1a40) in the remote environment).
The visible errors in the remote environment say that the files in '/home/runner/work/qiskit-ignis/qiskit-ignis/.tox/docs/lib/python3.8/site-packages/qiskit/quantum_info/' still have the error by using "np.int".

I'm not sure where to change...
Could you help me with the following error?

> ERROR: InvocationError for command /home/runner/work/qiskit-ignis/qiskit-ignis/.tox/docs/bin/sphinx-build -b html -W docs/ docs/_build/html (exited with code 2)
___________________________________ summary ____________________________________
ERROR:   docs: commands failed
Error: Process completed with exit code 1.

BOBO1997 commented:
And thank you for noticing me the issue #547. Actually, I am working together with Dr. Raymond and the motivation of this pull request is to somehow conduct the error mitigation in a bigger qubit size results than before (ideally ~50 qubits and more, but it is not yet possible).

For the approach in [Mooney et al.](https://arxiv.org/abs/2101.08946), they cut off the small elements of probability vector in every step of applying the inverse matrix to i-th measured bit.
In this sense, I think their process can scale in O(n * shots) by choosing a correct threshold.
However, as you pointed out, their heuristic method does not have theoretical guarantees and it seems to work correctly only for the measurements of quantum states with a small number of superposition such as GHZ states...
For this point, it is interesting to think about more efficient approaches with some theoretical guarantees.

For other approaches, I've found that we might try [the methods by Bravyi et al.](https://arxiv.org/abs/1612.02058), which are already prepared by [Mitiq](https://mitiq.readthedocs.io/en/stable/readme.html).

yaelbh reviewed (approved):
Thanks again, it's very nice!   A small final comment: complexity of full tensor (i.e., bitwise) is indeed `n*2^n`. Since we have `mit_pattern`, the complexity is `m*2^{t+n}`, where `m` is the number of sets in `mit_pattern`, and `t` is the size of the largest set. Special cases are bit-by-bit (`m=n, t=1`) and full calibration (`m=1, t=n`).

BOBO1997 commented:
You're right, thank you for your comments!
I fixed this part in the documents and made new commits.
I guess these commits will complete the main task for this pull request.
